task/sensor/src/main.rs

```diff
@@ -51,25 +57,82 @@ impl idl::InOrderSensorImpl for ServerImpl {
 
     fn get_reading(
         &mut self,
-        _: &RecvMessage,
+        msg: &RecvMessage,
         id: SensorId,
     ) -> Result<Reading, RequestError<SensorError>> {
+        self.get_raw_reading(msg, id)
+            .map_err(|e| e.map_runtime(SensorError::from))
+            .and_then(|(value, timestamp)| match value {
+                Ok(value) => Ok(Reading { value, timestamp }),
+                Err(e) => Err(<SensorError as From<_>>::from(e).into()),
+            })
+    }
+
+    fn get_raw_reading(
+        &mut self,
+        _: &RecvMessage,
+        id: SensorId,
+    ) -> Result<(Result<f32, NoData>, u64), RequestError<SensorApiError>> {
+        let index = id.0 as usize;
+
+        if index < NUM_SENSORS {
```

Comment:
Nit - I'm sure the compiler can optimize out the bounds check in `self.last_reading[index]` given this `if`, but I think this guarantees it and is slightly clearer:

```rust
let last_reading = self
    .last_reading
    .get(index)
    .ok_or(SensorApiError::InvalidSensor)?;

match last_reading {
    // ...
}
```

(same nit in more functions below this, which makes me now realize that this method is copying their style - sorry)

Reply:
I went slightly overboard in a1151980df9261cebd0e555d4bb83380c8735489, adding a new class to let us index directly by `SensorId` and refactoring `last_reading` into a standalone function (which is basically the code that you posted).